File: Add-Two-Numbers.py
# -*- coding: utf-8 -*-
# Definition for singly-linked list.


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    # @return a ListNode
    def addTwoNumbers(self, l1, l2):
        l1_list, l2_list = [], []
        while True:
            l1_list.append(l1.val)
            if not l1.next:
                break
            l1 = l1.next
        while True:
            l2_list.append(l2.val)
            if not l2.next:
                break
            l2 = l2.next
        l1_num, l2_num = 0, 0
        for i in l1_list[::-1]:
            l1_num = l1_num * 10 + i
        for i in l2_list[::-1]:
            l2_num = l2_num * 10 + i
        ret_num = l1_num + l2_num
        ret_l = ListNode(ret_num % 10)
        tmp = ret_l
        for i in str(ret_num)[-2::-1]:
            tmp.next = ListNode(int(i))
            tmp = tmp.next
        return ret_l


if __name__ == '__main__':
    solution = Solution()
    l1, l2 = ListNode(5), ListNode(5)
    _ = solution.addTwoNumbers(l1, l2)
    while True:
        print _.val,
        if not _.next:
            break
        _ = _.next

File: src/algorithms/cpp/Add_Two_Numbers.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int left = 0;
        ListNode *pre = NULL, *node = NULL, *head = l1;
        while (l1 && l2) {
            int tmp = l1->val + l2->val + left;
            pre = l1;
            l1->val = tmp % 10, left = tmp / 10;
            l1 = l1->next, l2 = l2->next;
        }
        if (l1) {
            node = l1;
        }
        else {
            node = l2;
        }
        while (node) {
            int tmp = node->val + left;
            node->val = tmp % 10, left = tmp/ 10;
            pre->next = node;
            pre = node;
            node = node->next;
        }
        if (left) {
            ListNode *new_node = new ListNode(left);
            pre->next = new_node;
        }
        return head;
    }
};

int main() {
    return 0;
}
